File: Domains/Base/Database/Domain/Base/Query/Instanceof.cs
// --------------------------------------------------------------------------------------------------------------------
// <copyright file="Instanceof.cs" company="Allors bvba">
//   Copyright 2002-2017 Allors bvba.
//
// Dual Licensed under
//   a) the General Public Licence v3 (GPL)
//   b) the Allors License
//
// The GPL License is included in the file gpl.txt.
// The Allors License is an addendum to your contract.
//
// Allors Applications is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// For more information visit http://www.allors.com/legal
// </copyright>
// --------------------------------------------------------------------------------------------------------------------

namespace Allors.Domain.Query
{
    using Allors.Meta;

    public class Instanceof : Predicate
    {
        public AssociationType AssociationType { get; set; }

        public RoleType RoleType { get; set; }

        public IComposite ObjectType { get; set; }

        public override void Build(ISession session, ICompositePredicate compositePredicate)
        {
            if (this.AssociationType != null)
            {
                compositePredicate.AddInstanceof(this.AssociationType, this.ObjectType);
            }
            else if (this.RoleType != null)
            {
                compositePredicate.AddInstanceof(this.RoleType, this.ObjectType);
            }
            else
            {
                compositePredicate.AddInstanceof(this.ObjectType);
            }
        }
    }
}

File: Domains/Base/Workspace/Typescript/Angular/src/allors/domain/base/database/query/Instanceof.ts
import { ObjectType, AssociationType, RoleType } from '../../../../meta';
import { Predicate } from './Predicate';

export class Instanceof implements Predicate {
  associationType: AssociationType;
  roleType: RoleType;
  objectType: ObjectType;

  constructor(fields?: Partial<Instanceof>) {
    Object.assign(this, fields);
  }

  toJSON(): any {
    return {
      _T: 'Instanceof',
      at: this.associationType ? this.associationType.id : undefined,
      rt: this.roleType.id ? this.roleType.id : undefined,
      ot: this.objectType.id,
    };
  }
}
